### PR TITLE
refactoring Cov2D codes

### PR DESCRIPTION
**Pull request by junchaoxia**
This PR  includes the enhancements requested in issues #52 and #34 for improving the Cov2D codes.

### CODE REVIEW

File: examples/denoise_covar2d.py [hunk 1]
```diff
@@ -21,14 +20,14 @@
 from aspire.image import Image
 from aspire.utils.matrix import anorm
 from aspire.utils.matlab_compat import randn
-from aspire.denoise.covar2d import RotCov2D
-from aspire.denoise.covar2d_ctf import Cov2DCTF
+from aspire.estimation.covar2d import RotCov2D
 from aspire.utils.blk_diag_func import blk_diag_add, blk_diag_mult, blk_diag_norm
 
-logger = logging.getLogger(__name__)
+logger = logging.getLogger('aspire')
```
Comment (janden) [inline examples/denoise_covar2d.py:26]: Why this change?

File: examples/denoise_covar2d.py [hunk 2]
```diff
@@ -75,23 +76,26 @@
...
 h_idx = np.array([filters.index(f) for f in sim.filters])
 
 # Evaluate CTF in the 8X8 FB basis
-h_ctf_fb = [radial_filter2fb_mat(filt.evaluate_k, ffbbasis) for filt in filters]
+h_ctf_fb = [radial_filter2fb_mat(filt.evaluate, ffbbasis) for filt in filters]
 
 # Apply the CTF to the clean images.
+logger.info('Apply CTF filters to clean images.')
 imgs_ctf_clean = Image(sim.eval_filters(imgs_clean))
 sim.cache(imgs_ctf_clean)
```
Comment (janden) [inline examples/denoise_covar2d.py:92]: What does this do? Is it necessary?

File: examples/denoise_covar2d.py [hunk 3]
```diff
@@ -75,23 +76,26 @@
...
 
 # imgs_ctf_clean is an Image object. Convert to numpy array for subsequent statements
 imgs_ctf_clean = imgs_ctf_clean.asnumpy()
```
Comment (janden) [inline examples/denoise_covar2d.py:95]: This shouldn't be necessary. We should be able to add noise (or any stack of images) to an Image object.

File: examples/denoise_covar2d.py [hunk 4]
```diff
@@ -126,17 +132,17 @@
 # estimates. For the covariance estimation, the additional information of
 # the estimated mean and the variance of the noise are needed. Again, the
 # covariance matrix estimate is provided in block diagonal form.
-cov2dctf = Cov2DCTF(sim, ffbbasis)
-mean_coeff_est = cov2dctf.get_mean_ctf(coeff_noise, h_ctf_fb, h_idx)
-covar_coeff_est = cov2dctf.get_covar_ctf(coeff_noise, h_ctf_fb, h_idx, mean_coeff_est, noise_var=noise_var)
+mean_coeff_est = cov2d.get_mean_ctf(coeff_noise, h_ctf_fb, h_idx)
```
Comment (janden) [inline examples/denoise_covar2d.py]: As discussed, we shouldn't have separate functions (at least exposed publicly) for with and without CTF. The CTF case can be reduced to the non-CTF case by assigning all images to the same filter and making that filter equal to the identity.

File: examples/denoise_covar2d.py [hunk 5]
```diff
@@ -126,17 +132,17 @@
...
 
 # Estimate the Fourier-Bessel coefficients of the underlying images using a
 # Wiener filter. This Wiener filter is calculated from the estimated mean,
 # covariance, and the variance of the noise. The resulting estimator has
 # the lowest expected mean square error out of all linear estimators.
-coeff_est = cov2dctf.get_cwf_coeffs(coeff_noise, h_ctf_fb, h_idx,
-                                    mean_coeff=mean_coeff_est,
-                                    covar_coeff=covar_coeff_est, noise_var=noise_var)
+logger.info('Get the CWF coefficients of noising images.')
+coeff_est = cov2d.get_cwf_coeffs(coeff_noise, h_ctf_fb, h_idx,
```
Comment (janden) [inline examples/denoise_covar2d.py:147]: This shouldn't be a function of the `RotCov2D` object, since it has nothing to do with estimating the 2D covariance. Maybe it would be better if this were a standalone function?

File: examples/denoise_covar2d.py [hunk 6]
```diff
@@ -154,8 +160,8 @@
 # Calculate the normalized RMSE of the estimated images.
 nrmse_ims = anorm(imgs_est-imgs_clean)/anorm(imgs_clean)
 
-print(f'Deviation of the noisy mean estimate: {diff_mean}')
-print(f'Deviation of the noisy covariance estimate: {diff_covar}')
-print(f'Estimated images normalized RMSE:{nrmse_ims}')
+logger.info(f'Deviation of the noisy mean estimate: {diff_mean}')
+logger.info(f'Deviation of the noisy covariance estimate: {diff_covar}')
+logger.info(f'Estimated images normalized RMSE:{nrmse_ims}')
```
Comment (janden) [inline examples/denoise_covar2d.py]: Space between colon and `{nrmse_ims}`.

File: src/aspire/utils/blk_diag_func.py [hunk 7]
```diff
@@ -165,7 +165,7 @@ def blk_diag_isnumeric(x):
 def radial_filter2fb_mat(h_fun, fbasis):
     n_r = fbasis.sz[0]
     k_vals, wts = lgwt(n_r, 0, 0.5)
-    h_vals = h_fun(k_vals)
+    h_vals = h_fun(k_vals*2 * np.pi)
```
Comment (janden) [inline src/aspire/utils/blk_diag_func.py]: Why is this necessary? Also, space around the first asterisk.

File: src/aspire/estimation/covar2d.py [hunk 8]
```diff
@@ -207,3 +292,7 @@ def get_cwf_coeffs(self, coeffs, ctf_fb, ctf_idx, mean_coeff=None, covar_coeff=N
             coeffs_est[:, ctf_idx == k] = coeff_est_k
 
         return coeffs_est
+
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Remove all these extra lines.

File: src/aspire/estimation/covar2d.py [hunk 9]
```diff
@@ -9,25 +9,107 @@
 
 from aspire.utils.matrix import shrink_covar
 from aspire.utils.optimize import conj_grad
-from aspire.denoise.covar2d import RotCov2D
-from aspire.utils import ensure
 
+from aspire.utils import ensure
```
Comment (janden) [inline src/aspire/estimation/covar2d.py:10]: Why move this down a line?

File: src/aspire/estimation/covar2d.py [hunk 10]
```diff
@@ -9,25 +9,107 @@
...
 
 logger = logging.getLogger(__name__)
 
 
-class Cov2DCTF(RotCov2D):
+class RotCov2D:
     """
-    Define a derived class for denoising 2D images using CTF information and the Covariance Wiener Filtering (CWF)
-    Cov2D method described in
+    Define a class for denoising 2D images using the Cov2D method with CTF information described in
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: This needs to be updated. The class longer does denoising, but estimates the 2D covariance.

File: src/aspire/estimation/covar2d.py [hunk 11]
```diff
@@ -9,25 +9,107 @@
...
     J. Struct. Biol. 195, 27-81 (2016). DOI: 10.1016/j.jsb.2016.04.013
     """
 
+    def __init__(self, basis, as_type='single'):
+        """
+        constructor of an object for 2D covariance analysis
+        """
+        self.basis = basis
+        ensure(basis.d == 2, 'Only two-dimensional basis functions are needed.')
+        self.as_type = as_type
```
Comment (janden) [inline src/aspire/estimation/covar2d.py:29]: Why does this have to be specified here? Shouldn't the type be inferred from the input (`coeffs`)? Also, it seems like it's only used in `get_mean`.

File: src/aspire/estimation/covar2d.py [hunk 12]
```diff
@@ -9,25 +9,107 @@
...
+
+    def get_mean(self, coeffs):
+        """
+        Calculate the mean vector from the expansion coefficients of 2D images without CTF information.
+
+        :param coeffs: A coefficient vector (or an array of coefficient vectors) to be averaged.
+        :return: The mean value vector for all images.
+        """
+        if coeffs is None:
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Why would a user send in `None` as the coefficients?

File: src/aspire/estimation/covar2d.py [hunk 13]
```diff
@@ -9,25 +9,107 @@
...
+            raise RuntimeError('The coefficients need to be calculated first!')
+
+        mask = self.basis._indices["ells"] == 0
+        mean_coeff = np.zeros((self.basis.basis_count, 1), dtype=self.as_type)
```
Comment (janden) [inline src/aspire/estimation/covar2d.py:42]: Why does `mean_coeff` have two dimensions? It's just a vector.

File: src/aspire/estimation/covar2d.py [hunk 14]
```diff
@@ -9,25 +9,107 @@
...
+    def get_covar(self, coeffs, mean_coeff=None,  do_refl=True):
+        """
+        Calculate the covariance matrix from the expansion coefficients without CTF information.
+
+        :param coeffs: A coefficient vector (or an array of coefficient vectors) calculated from 2D images.
+        :param mean_coeff: The mean vector calculated from the `coeffs`.
+        :param do_refl: If true, enforce invariance to reflection (default false).
+        :return: The covariance matrix of coefficients for all images.
+        """
+        if coeffs is None:
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Again, why needed?

File: src/aspire/estimation/covar2d.py [hunk 15]
```diff
@@ -9,25 +9,107 @@
...
+            raise RuntimeError('The coefficients need to be calculated first!')
+        if mean_coeff is None:
```
Comment (janden) [inline src/aspire/estimation/covar2d.py:58]: Same here.

File: src/aspire/estimation/covar2d.py [hunk 16]
```diff
@@ -9,25 +9,107 @@
...
+            mean_coeff = self.get_mean(coeffs)
+
+        covar_coeff = []
+        ind = 0
+        ell = 0
+        mask =  self.basis._indices["ells"] == ell
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Remove double space after equals sign.

File: src/aspire/estimation/covar2d.py [hunk 17]
```diff
@@ -9,25 +9,107 @@
...
+            mask = self.basis._indices["ells"] == ell
+            mask_pos = [mask[i] and (self.basis._indices['sgns'][i] == +1) for i in range(len(mask))]
+            mask_neg = [mask[i] and (self.basis._indices['sgns'][i] == -1) for i in range(len(mask))]
+            covar_ell_diag = np.array(coeffs[mask_pos, :] @ coeffs[mask_pos, :].T +
+                coeffs[mask_neg, :] @ coeffs[mask_neg, :].T) / (2 * np.size(coeffs, 1))
+
+            if do_refl:
+                covar_coeff.append(covar_ell_diag)
+                covar_coeff.append(covar_ell_diag)
+                ind = ind+2
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Space around plus sign.

File: src/aspire/estimation/covar2d.py [hunk 18]
```diff
@@ -9,25 +9,107 @@
...
+                hsize = np.size(covar_ell_diag, 0)
+                covar_coeff_blk = np.zeros((2*hsize, 2*hsize))
+
+                fsize = np.size(covar_coeff_blk, 0)
+                covar_coeff_blk[0:hsize, 0:hsize] = covar_ell_diag[0:hsize, 0:hsize]
+                covar_coeff_blk[hsize:fsize, hsize:fsize] = covar_ell_diag[0:hsize, 0:hsize]
+                covar_coeff_blk[0:hsize, hsize:fsize] = covar_ell_off[0:hsize, 0:hsize]
+                covar_coeff_blk[hsize:fsize, 0:hsize] = covar_ell_off.T[0:hsize, 0:hsize]
+                covar_coeff.append(covar_coeff_blk)
+                ind = ind+1
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Space around plus sign.

File: src/aspire/estimation/covar2d.py [hunk 19]
```diff
@@ -9,25 +9,107 @@
...
+
+            if do_refl:
+                covar_coeff.append(covar_ell_diag)
+                covar_coeff.append(covar_ell_diag)
+                ind = ind+2
+            else:
+                covar_ell_off = np.array((coeffs[mask_pos, :] @ coeffs[mask_neg, :].T / np.size(coeffs, 1) -
+                                 coeffs[mask_neg, :] @ coeffs[mask_pos, :].T)/(2*np.size(coeffs, 1)))
+                hsize = np.size(covar_ell_diag, 0)
+                covar_coeff_blk = np.zeros((2*hsize, 2*hsize))
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Space around asterisk.

File: src/aspire/estimation/covar2d.py [hunk 20]
```diff
@@ -9,25 +9,107 @@
...
+            covar_ell_diag = np.array(coeffs[mask_pos, :] @ coeffs[mask_pos, :].T +
+                coeffs[mask_neg, :] @ coeffs[mask_neg, :].T) / (2 * np.size(coeffs, 1))
+
+            if do_refl:
+                covar_coeff.append(covar_ell_diag)
+                covar_coeff.append(covar_ell_diag)
+                ind = ind+2
+            else:
+                covar_ell_off = np.array((coeffs[mask_pos, :] @ coeffs[mask_neg, :].T / np.size(coeffs, 1) -
+                                 coeffs[mask_neg, :] @ coeffs[mask_pos, :].T)/(2*np.size(coeffs, 1)))
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Space around division sign and asterisk.

File: src/aspire/estimation/covar2d.py [hunk 21]
```diff
@@ -9,25 +9,107 @@
...
+                covar_coeff_blk[0:hsize, 0:hsize] = covar_ell_diag[0:hsize, 0:hsize]
+                covar_coeff_blk[hsize:fsize, hsize:fsize] = covar_ell_diag[0:hsize, 0:hsize]
+                covar_coeff_blk[0:hsize, hsize:fsize] = covar_ell_off[0:hsize, 0:hsize]
+                covar_coeff_blk[hsize:fsize, 0:hsize] = covar_ell_off.T[0:hsize, 0:hsize]
+                covar_coeff.append(covar_coeff_blk)
+                ind = ind+1
+
+        return covar_coeff
+
+    def get_coeffs(self, src):
```
Comment (janden) [inline src/aspire/estimation/covar2d.py]: Why is this function needed here? Shouldn't this be done elsewhere?

**Review by janden — CHANGES_REQUESTED** [review]
Some things to discuss here, but overall, it looks good.

File: examples/denoise_covar2d.py (hunk 1 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py:26]: We want to use a more consistent format for logging as the scripts in  `src/aspire/commands/`.

File: examples/denoise_covar2d.py (hunk 2 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py:92]: Currently the simulation class only generates images from Gaussian blob volume. This cache operation will replace the images from Gaussian blob volume with the images from experimental volume. This is only for the consistency with the replaced Gaussian volume in  a previous operation, `sim.vols = vols` in line 71.

File: examples/denoise_covar2d.py (hunk 3 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py:95]: Agree. We should have simpler operation when the` Image` and `ImageSource` classes are finalized. The example is doing this way to make sure the added noises are consistent with the Matlab example.

File: examples/denoise_covar2d.py (hunk 4 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py]: I will try to combine the non-CTF and CTF methods in the same class.

File: examples/denoise_covar2d.py (hunk 5 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py:147]: You are right. Maybe we can put it in the new `Denoise` or `Coefficient` class.

File: examples/denoise_covar2d.py (hunk 6 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py]: done

File: src/aspire/utils/blk_diag_func.py (hunk 7 above)
Comment (junchaoxia) [reply src/aspire/utils/blk_diag_func.py]: This function transfers the CTF filter functions in k space into FB coefficients so that it can operate together using block diagonal matrix.

File: src/aspire/estimation/covar2d.py (hunk 8 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done

File: src/aspire/estimation/covar2d.py (hunk 9 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py:10]: done.

File: src/aspire/estimation/covar2d.py (hunk 10 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done.

File: src/aspire/estimation/covar2d.py (hunk 11 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py:29]: This initializes a default data type for calculating mean and covariance,  and also sets the data type for block diagonal matrix. Add the statement to get the inferred value from `coeffs`.

File: src/aspire/estimation/covar2d.py (hunk 12 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: Initially this is to make sure that a non-empty `coeffs` array is passed. Now change to check the size.

File: src/aspire/estimation/covar2d.py (hunk 13 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py:42]: Just make a consistent number of dimensions with `coeffs` for future operations without dimension change.

File: src/aspire/estimation/covar2d.py (hunk 14 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: Same reason. Now change to check the size.

File: src/aspire/estimation/covar2d.py (hunk 15 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py:58]: It allows to call the function directly for computing the covariance without calculating mean value first.

File: src/aspire/estimation/covar2d.py (hunk 16 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done.

File: src/aspire/estimation/covar2d.py (hunk 17 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done.

File: src/aspire/estimation/covar2d.py (hunk 18 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done.

File: src/aspire/estimation/covar2d.py (hunk 19 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done

File: src/aspire/estimation/covar2d.py (hunk 21 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: Not implemented. Removed it now. Probably put it in the  `Coefficient `class.

File: src/aspire/estimation/covar2d.py (hunk 20 above)
Comment (junchaoxia) [reply src/aspire/estimation/covar2d.py]: done.

File: examples/denoise_covar2d.py (hunk 4 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py]: I merged them by checking the CTF information exists or not. It looks that we don't need to apply the identity filter at all. The update is done in commit 454d85a.

File: examples/denoise_covar2d.py (hunk 1 above)
Comment (janden) [reply examples/denoise_covar2d.py:26]: Ok.

File: examples/denoise_covar2d.py (hunk 3 above)
Comment (janden) [reply examples/denoise_covar2d.py:95]: Ok. Let's make an issue.

File: examples/denoise_covar2d.py (hunk 4 above)
Comment (janden) [reply examples/denoise_covar2d.py]: Ok.

File: src/aspire/utils/blk_diag_func.py (hunk 7 above)
Comment (janden) [reply src/aspire/utils/blk_diag_func.py]: Yes, but why wasn't the `2 * np.pi` factor needed before?

File: examples/denoise_covar2d.py (hunk 3 above)
Comment (junchaoxia) [reply examples/denoise_covar2d.py:95]: Open the issue #61 for addressing this in future.

File: src/aspire/estimation/covar2d.py (hunk 11 above)
Comment (janden) [reply src/aspire/estimation/covar2d.py:29]: I still don't understand why this needs to be stored as an attribute. Couldn't `get_mean` and `get_covar` just use the data type of `coeffs`?

File: src/aspire/estimation/covar2d.py (hunk 12 above)
Comment (janden) [reply src/aspire/estimation/covar2d.py]: I don't think that's necessary. If you try to call a function without the required number of arguments, it already gives an error. Why would someone send in an array of size zero?

File: src/aspire/estimation/covar2d.py (hunk 13 above)
Comment (janden) [reply src/aspire/estimation/covar2d.py:42]: I don't think that's a good enough reason. You can add, subtract, etc. a 2D array and a 1D array (vector) in NumPy without any trouble. This just adds unnecessary complexity.

File: src/aspire/estimation/covar2d.py (hunk 14 above)
Comment (janden) [reply src/aspire/estimation/covar2d.py]: Same here. It's not necessary. Remove.

File: src/aspire/estimation/covar2d.py (hunk 15 above)
Comment (janden) [reply src/aspire/estimation/covar2d.py:58]: Ok sure, but then a saner default would be to use a zero vector for the mean, I think.

File: examples/denoise_covar2d.py (hunk 2 above)
Comment (janden) [reply examples/denoise_covar2d.py:92]: Ok. That's a bad way to use a simulation. As discussed yesterday, we need a new type of `ImageSource` (simply backed by a stack of images that we provide) for this type of use case. A `cache()` function should never accept a stack of images as input, it should simply cache the *existing* images in that `ImageSource`. Let's make an issue for this.

File: examples/denoise_covar2d.py (hunk 5 above)
Comment (janden) [reply examples/denoise_covar2d.py:147]: It should be in the denoiser, I think.

**Review by janden — COMMENTED** [review]
Few more things to look at.

File: src/aspire/estimation/covar2d.py [hunk 22]
```diff
@@ -148,19 +229,20 @@ def apply(A, x):
...
         """
         b_out = b
-        for ell in range(0, b.size()):
+        for ell in range(0, len(b)):
             b_ell = b[ell]
             p = np.size(b_ell, 1)
             S = sqrtm(b_noise[ell])
             # from Matlab b_ell = S \ b_ell /S
-            b_ell = np.divide(solve(S, b_ell), S)
+            b_ell = solve(S, b_ell) @ inv(S)
```
Comment (janden) [inline src/aspire/estimation/covar2d.py:245]: This is not very efficient (or stable). Doing something like `solve(S.T, solve(S, b_ell).T).T` is the right way to do this.
